p2p/examples/dynamic-firewall.rs

```diff
@@ -0,0 +1,282 @@
+//! T1:
+//! ```sh
+//! Local Peer Id: 12D3KooWRgJG7no2snYqGM8jwABTs952KFb6GHghzz1vRw7T5Cmd
+//! ```
+//! T2:
+//! ```sh
+//! Local Peer Id: 12D3KooWG364edSdRCv5LG5adzsAdE3vafYqc5rGHCYpvT7dmHPa
+//! ```
+//!
+//! Because mDNS is enabled in this example, peers aromatically learn the listening addresses of other peers
+//! in the same network, that is used to dial this peer.
+//!
+//!
+//! In T1 run
+//! ```sh
+//! -p 12D3KooWG364edSdRCv5LG5adzsAdE3vafYqc5rGHCYpvT7dmHPa -m "test message"
+//! ```
+//! the id being the one that was printed in T2.
+//!
+//! In the second terminal, it will ask for the firewall rules:
+//! T2:
+//! ```sh
+//! Peer 12D3KooWRgJG7no2snYqGM8jwABTs952KFb6GHghzz1vRw7T5Cmd connected. Allow requests from them?: (yes/no/ask/ping)
+//! > ask
+//!
+//! # Ask for individual approval due to rule "Ask"
+//! Received Request with type Message from peer 12D3KooWRgJG7no2snYqGM8jwABTs952KFb6GHghzz1vRw7T5Cmd. Permit?: (yes/no)
+//! > yes
+//!
+//! # The actual message is shown
+//! Received Message from peer 12D3KooWRgJG7no2snYqGM8jwABTs952KFb6GHghzz1vRw7T5Cmd:
+//! "test message".
+//! > test response
+//! ```
+//!
+//! The response will then be printed in T1:
+//! ```sh
+//! Response: test response
+//! ```
+//! Note: While waiting for a response T1 is blocking.
+
+use futures::{channel::mpsc, FutureExt, StreamExt};
+use p2p::{
+    firewall::{
+        permissions::{FirewallPermission, PermissionValue, RequestPermissions, VariantPermission},
+        FirewallRequest, FirewallRules, Rule,
+    },
+    ChannelSinkConfig, EventChannel, PeerId, ReceiveRequest, RequestDirection, StrongholdP2p,
+};
+use regex::Regex;
+use serde::{Deserialize, Serialize};
+use std::{borrow::Borrow, error::Error, marker::PhantomData, str::FromStr, time::Duration};
+use tokio::io::{stdin, AsyncBufReadExt, BufReader, Lines, Stdin};
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, RequestPermissions)]
+enum Request {
+    Ping,
+    Message(String),
+}
+
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
+enum Response {
+    Pong,
+    Message(String),
+}
+
+// Permit only messages of type "Ping". See `p2p::firewall:permissions` module for more info.
+fn allow_only_ping(request: &RequestPermission) -> bool {
+    let allowed_variant = FirewallPermission::none().add_permissions([&RequestPermission::Ping.permission()]);
+    allowed_variant.permits(&request.permission())
+}
+
+// Handle user input through stdin.
+// Expect either of:
+// - `-p <peer-id>`: Send a ping
+// - `-p <peer-id> -m <message>`: Send a message
+async fn on_user_input(
+    network: &mut StrongholdP2p<Request, Response, RequestPermission>,
+    input: String,
+) -> Result<(), Box<dyn Error>> {
+    let peer_regex = "-p\\s+(?P<target>[[:alnum:]]{32,64})";
+    let peer: PeerId = match Regex::new(peer_regex).expect("Valid regex").captures(&input) {
+        Some(capture) => {
+            let target = capture.name("target").expect("Target was captured.");
```

Comment:
```suggestion
            let target = capture.name("target").expect("Target was captured.");
```
minor: it's rather "capture group not found"

Reply:
I personally like to use the message in `expect` to describe the expected invariant (= why to we expect it to be safe to unwrap here) rather than an error message that is just stating the obvious. Ideally a library/ application should never unwrap values unless they know that the value is `Ok`/ `Some`, so the message should never be shown to the user.

However, I agree that in this case the wording is not ideal. What do you think about something along the lines of 
`.expect("regex should only match strings with a 'target' capture group.").`
Alternatively we could also just use `unwrap` here and add a comment on why it is safe to unwrap.

Reply:
> "regex should only match strings with a 'target' capture group."

I think that's a good choice.

Reply:
Done in 05df78f.